Add more weapon burst controls and simplify RA Vulcan setup

reaperrr opened the pull request:
Added more configurability for burst behavior to `WeaponInfo`.

Closes #13220.
Fixes #18312.

reaperrr committed:
Remove attack notification from SwallowActor

This
- ignored potential projectile launch delays,
- should have been redundant, because Armament already
  does all this (and in the right way).

For this to work for actors like sandworms,
Reports, notifications and recoil are no longer
requiring a projectile.

AttackGarrisoned still needs this, but INotifyAttack
traits are now cached.

reaperrr committed:
Add more weapon burst controls

This adds fields that allow weapons to
fire all bursts at a static position,
or firing all of them even if the target dies.

reaperrr committed:
Simplify RA Vulcan setup

Uses new Burst-related fields to get rid
of all delayed duplicate warheads on
RA Vulcan weapon, while fully maintaining its
bleed behaviour.

reaperrr committed:
Increase Pillbox muzzle anim length

Length of muzzle anim now roughly matches the length
of the sound and impact animations.

pchote commented on OpenRA.Game/GameRules/WeaponInfo.cs:
```diff
@@ -64,6 +64,8 @@ public WarheadArgs(WarheadArgs args)
 		public WarheadArgs() { }
 	}
 
+	public enum BurstMode { Individual, AllOnInitialTargetPosition }
```
How about `Fixed` or `Static` as a less-of-a-mouthful name than `AllOnInitialTargetPosition`?

reaperrr replied:
Ok. I'm not entirely sure if `Static` alone would be enough though.

The other 2 modes I have in mind for a follow-up are
(1) follows target like `Individual`, but cancels burst (drops any remaining bursts to zero) once target becomes invalid in some way
(2) always depletes all bursts no matter what, but follows target instead of shooting at fixed position as long as target is alive and moving.

Suggestions on how to call these, to avoid having to rename the two introduced here later?

Would something like
`Flexible` (currently `Individual`)
`CancelOnTargetDeath` (1)
`DepleteOnTargetDeath` (2)
`Static`
be ok?
Not sure how to shorten the middle ones without making it hard to differentiate them...

pchote replied:
I've put less thought into this than you have, but given those use cases my first thought would to be use separate flags:
```csharp
[Desc("Update the firing position to follow targets if they move between bursts?")]
public readonly bool BurstTracking = true;

[Desc("Allow the weapon to cancel its target between bursts if requested by the player or the target dies?")]
public readonly bool BurstInterruptable = true;

[Desc("Reset the burst counter to 0 when switching targets?")]
public readonly bool BurstResetOnTargetChange = false;

```

It would be ideal if we could also fix `FireDelay` to update the target position if `BurstTracking = true` / `BurstMode = Individual`.

PunkPun commented:
> Added a `BurstMode` enum to `WeaponInfo`, with the `AllOnInitialTargetPosition` option allowing to replicate RA's `Vulcan` behavior without the yaml-heavy work-around of using tons of delayed warheads.

Nuke weapons also have a ton of delayed warheads

reaperrr committed:
Refactor FireBarrel and split off CreateProjectileArgs

Deferring all calculations and state look-ups that
are part of creating the projectile args to the time
the projectile is actually needed can avoid wasting
perf (if creator dies before barrel is launched) and ensures
that all values are up to date.

Also performed some cleanup: facing appears to have been
unused for some time already.

reaperrr commented:
> Nuke weapons also have a ton of delayed warheads

This cannot be applied to nukes, they're a different beast.
Each burst would trigger all warheads, including the explosion effect.

It only works for the Vulcan because the warheads at each delay were otherwise 100% identical, and triggering the piffpiff effect each time is fine and actually intended.

reaperrr commented:
> I've put less thought into this than you have, but given those use cases my first thought would to be use separate flags:

Done, although I left out `BurstResetOnTargetChange` for now since that would complicate things further and isn't needed for the Vulcan refactor. I'll probably file a follow-up once this is merged.

> It would be ideal if we could also fix `FireDelay` to update the target position if `BurstTracking = true`

Done.

reaperrr commented:
Rebased and fixes #18312, too.

Smittytron commented:
LGTM, fix for #18312 confirmed.

pchote commented on OpenRA.Mods.Common/Traits/Armament.cs:
```diff
@@ -251,40 +254,111 @@ protected virtual bool CanFire(Actor self, Target target)
+				for (var i = Burst; i > 0; i--)
+				{
+					var barrel = Barrels[Burst % Barrels.Length];
+					FireBarrel(self, target, barrel, delay, firstBurst);
```
We can remove `firstBurst` by using `FireBarrel(self, target, barrel, delay, i == 0);` instead.

pchote commented on OpenRA.Mods.Common/Traits/Armament.cs:
```diff
@@ -251,40 +254,111 @@ protected virtual bool CanFire(Actor self, Target target)
+		//
+		// 1) BurstTracking = true + BurstInterruptable = true: the default behavior, namely
+		// 		- firing each burst at the targets' position at the time of firing
+		// 		- using remaining bursts on the next target if target dies before bursts are depleted
```
target dies, or the player stops / requests a new target(?)

pchote commented on OpenRA.Mods.Common/Traits/Armament.cs:
```diff
@@ -251,40 +254,111 @@ protected virtual bool CanFire(Actor self, Target target)
+				var firstBurst = true;
+				for (var i = Burst; i > 0; i--)
+				{
+					var barrel = Barrels[Burst % Barrels.Length];
```
Why don't we need to update `currentBarrel` here?

reaperrr replied:
`currentBarrel` is only used for weapons with Burst == 1 to toggle the LocalOffsets.

Bulk scheduling only makes sense and therefore is only enabled/supported for Burst > 1, so there's no reason to update it.

pchote commented:
Do we understand what broke #18312 and how this fixes it? My main worry is if we don't then there may be other weapons that we haven't yet realized are broken.

pchote commented on OpenRA.Mods.Common/Traits/Attack/AttackGarrisoned.cs:
```diff
@@ -185,9 +184,6 @@ public override void DoAttack(Actor self, Target target)
 				}
-
-				foreach (var npa in self.TraitsImplementing<INotifyAttack>())
-					npa.Attacking(self, target, a, barrel);
```
Camo pillboxes no longer uncloak while attacking, so the implementation in `Armament` is somehow not covering this case!

pchote replied:
I guess that `Armament` will be notifying the passenger, while this is needed to notify the pillbox. Adding this back to just `AttackGarrisoned` should hopefully be sufficient.

reaperrr replied:
Fixed.

pchote reviewed (commented):
A few initial thoughts. I stopped testing once I discovered the camo pillbox issue because I don't know how much reworking that will require to fix.

Smittytron commented:
> Do we understand what broke #18312 and how this fixes it? My main worry is if we don't then there may be other weapons that we haven't yet realized are broken.

Looks to me like #18068. If you test with combat geometry only the first of six shots hits. By merging the rules for the other shots into the first one this PR fixes the problem.

reaperrr commented:
> Do we understand what broke #18312 and how this fixes it? My main worry is if we don't then there may be other weapons that we haven't yet realized are broken.

Yes, I forgot to properly update the ValidTargets on the delayed duplicate warheads. With the new setup there's only the original damage warhead, which doesn't have that issue.

reaperrr commented:
Rebased.

pchote commented on OpenRA.Mods.Common/Traits/Armament.cs:
```diff
@@ -251,40 +254,109 @@ protected virtual bool CanFire(Actor self, Target target)
+				for (var i = Burst; i > 0; i--)
+				{
+					var barrel = Barrels[Burst % Barrels.Length];
+					FireBarrel(self, target, barrel, delay, i == 0);
```
`i` is never equal to 0, so pillboxes (and other actors with weapons that rely on `StartBurstReport`) never play their firing audio. `i == Weapon.Burst` will do what you want, but see my overall comment above.

pchote commented on OpenRA.Mods.Common/Traits/Armament.cs:
```diff
@@ -251,40 +254,109 @@ protected virtual bool CanFire(Actor self, Target target)
+		{
+			if (!CanFire(self, target, true))
+				return null;
+			else if (!Weapon.BurstTracking && Weapon.BurstInterruptable
```
Style nit: Never use `else if` after a return - `if` is sufficient because we have already returned if the earlier branch was taken.

pchote reviewed (commented):
I'm sorry to say that I still have serious reservations about the `bulkScheduleBarrels` behaviour - having the same inner code being called from two different context makes it very difficult to mentally model the logic, and will almost certainly lead to unexpected bugs (I already have one below! I only spotted it because my IDE gave a warning about the condition being impossible).

Can we please change this so that all of the cases use the `!bulkScheduleBarrels` logic?

reaperrr commented:
> Can we please change this so that all of the cases use the !bulkScheduleBarrels logic?

The reason I haven't responded so far is that unless I'm missing something, this would pretty much mean throwing out the core part of this PR altogether, trying to find a completely different approach and (if I even succeed with the latter) rewriting the bulk behavior from scratch.
What I have here is the only working approach I was able to come up with though, so that would essentially boil down to giving up for now and deferring this to the undefined 'Future' again.

reaperrr commented:
I currently have no time/motivation for this (especially if the current approach isn't good enough and everything needs to be redone from scratch with a new conceptual approach), so to clean up the PR queue and Next milestone a bit I'm moving this to Future for now.
